network/forwarder/forwarder.go

```diff
@@ -0,0 +1,317 @@
+	}
+	if newForwarder.Log == nil {
+		return nil, maskAnyf(invalidConfigError, "logger must not be empty")
+	}
+	if newForwarder.StorageCollection == nil {
+		return nil, maskAnyf(invalidConfigError, "storage collection must not be empty")
+	}
+
+	// Settings.
+	if newForwarder.MaxSignals == 0 {
+		return nil, maskAnyf(invalidConfigError, "maximum signals must not be empty")
+	}
+
+	newForwarder.Log.Register(newForwarder.GetType())
+
+	return newForwarder, nil
+}
+
+// MustNew creates either a new default configured forwarder object, or panics.
+func MustNew() spec.Forwarder {
+	newForwarder, err := New(DefaultConfig())
+	if err != nil {
+		panic(err)
+	}
+
+	return newForwarder
+}
+
+type forwarder struct {
+	Config
+
+	ID   spec.ObjectID
+	Type spec.ObjectType
+}
+
+func (f *forwarder) Forward(CLG spec.CLG, networkPayload spec.NetworkPayload) error {
+	f.Log.WithTags(spec.Tags{C: nil, L: "D", O: f, V: 13}, "call Forward")
+
+	// This is the list of lookup functions which is executed seuqentially.
+	lookups := []func(CLG spec.CLG, networkPayload spec.NetworkPayload) ([]spec.NetworkPayload, error){
+		f.GetNetworkPayloads,
+		f.NewNetworkPayloads,
+	}
+
+	// Execute one lookup after another. As soon as we find some behaviour IDs, we
+	// use them to forward the given network payload.
+	var newNetworkPayloads []spec.NetworkPayload
+	var err error
+	for _, lookup := range lookups {
+		newNetworkPayloads, err = lookup(CLG, networkPayload)
+		if IsNetworkPayloadsNotFound(err) {
+			// There could no behaviour IDs be found by this lookup. Go on and try the
+			// next one.
+			continue
+		} else if err != nil {
+			return maskAny(err)
+		}
+
+		// The current lookup was successful. We do not need to execute any further
+		// lookup, but can go on with the behaviour IDs found.
+		break
+	}
+
+	// Forward the found network payloads to other CLGs by adding them to the
+	// queue so other processes can fetch them.
+	for _, np := range newNetworkPayloads {
+		networkPayloadKey := key.NewCLGKey("events:network-payload")
+		b, err := json.Marshal(np)
+		if err != nil {
+			return maskAny(err)
+		}
+		// TODO store asynchronuously
+		err = f.Storage().General().PushToList(networkPayloadKey, string(b))
+		if err != nil {
+			return maskAny(err)
+		}
+	}
+
+	return nil
+}
+
+func (f *forwarder) GetMaxSignals() int {
+	return f.MaxSignals
+}
+
+func (f *forwarder) GetNetworkPayloads(CLG spec.CLG, networkPayload spec.NetworkPayload) ([]spec.NetworkPayload, error) {
+	ctx := networkPayload.GetContext()
+
+	// Check if there are behaviour IDs known that we can use to forward the
+	// current network payload to.
+	var newBehaviourIDs []string
+	behaviourID, ok := ctx.GetBehaviourID()
+	if !ok {
+		return nil, maskAnyf(invalidBehaviourIDError, "must not be empty")
+	}
+	behaviourIDsKey := key.NewCLGKey("forward:configuration:behaviour-id:%s:behaviour-ids", behaviourID)
+	// TODO add closer instead of nil
```

Comment:
Instead of using the closer we should check if the set methods make sense to be used here. Further there should be a method to fetch all elements of a set at once. The forwarder limits the number of elements stored per CLG anyway, so walking the elements of a set incrementally is not necessary.